### PR TITLE
Belos: Register SolverFactory for ScalarType = float

### PR DESCRIPTION
**Pull request by jennloe**
This PR modifies solver factory registration routines so that the solver factory can compile and run with scalarType = float.  

@trilinos/belos 

## Testing

We copy existing solver factory tests to also run with scalarType = float.

### CODE REVIEW

File: packages/belos/src/BelosSolverFactory_Belos.hpp [hunk 1]
```diff
@@ -62,6 +62,14 @@ class BelosSolverFactory : public Impl::SolverFactoryParent<double,MultiVec<doub
     };
 };
 
+class BelosFloatSolverFactory : public Impl::SolverFactoryParent<float,MultiVec<float>,Operator<float>>
```
Comment (cgcgcg) [inline packages/belos/src/BelosSolverFactory_Belos.hpp:65]: Does this need to be guarded? The complex stuff has a `#ifdef HAVE_TEUCHOS_COMPLEX`.

Comment (jennloe) [reply packages/belos/src/BelosSolverFactory_Belos.hpp:65]: Possibly, but I am not sure what to guard it with.  When I wrapped the Factory.cpp test with `#ifdef Trilinos_ENABLE_FLOAT `, that test did not build, even though I had enabled that option.

Comment (jennloe) [reply packages/belos/src/BelosSolverFactory_Belos.hpp:65]: @hkthorn ??

Comment (hkthorn) [reply packages/belos/src/BelosSolverFactory_Belos.hpp:65]: I would say that "float" does not need to be guarded.  There are basic tests of LAPACK/BLAS/SDM in Teuchos that are based on float and are not guarded.  If the linear algebra stack supports float, then it should be fine.  Tpetra is controlling which scalar types it is enabling, like float, but this is not the only linear algebra that Belos can work with.

**Comment by jennloe** [comment]
@cgcgcg This is failing on a MueLu test.  Can you look?

**Review by cgcgcg — APPROVED** [review]

**Pull request merged**